src/Components/Components/src/ChangeDetection.cs

```diff
@@ -34,11 +34,13 @@ internal sealed class ChangeDetection
     // time. So we don't want a huge list of types to check (or would have to move to
     // a hashtable lookup, which is differently expensive). It's better not to include
     // uncommon types here even if they are known to be immutable.
+    // This logic assumes that no new System.TypeCode enum entries have been declared since 7.0, or at least that any new ones
+    // represent immutable types. If System.TypeCode changes, review this logic to ensure it is still correct.
+    // Supported immutable types : bool, byte, sbyte, short, ushort, int, uint, long, ulong, char, double,
+    //                             string, DateTime, decimal, Guid, Enum, EventCallBack, EventCallBack<>.
+    // For performance reasons, the following immutable types are not supported: IntPtr, UIntPtr, Type.
     private static bool IsKnownImmutableType(Type type)
-        => type.IsPrimitive
-        || type == typeof(string)
-        || type == typeof(DateTime)
-        || type == typeof(Type)
-        || type == typeof(decimal)
-        || type == typeof(Guid);
+        => Type.GetTypeCode(type) != TypeCode.Object
+        || type == typeof(Guid)
+        || typeof(IEventCallback).IsAssignableFrom(type);
```

Comment:
The logic here is wrong.

TypeCode does not cover `Type` (which is not covered in the test) and `IEventCallback` is not immutable.

I think perf-wise I do not have concerns, but the change is regressing the two scenarios above, which is not acceptable.

Reply:
> TypeCode does not cover Type (which is not covered in the test)

Unless I'm misreading, the point they are making in https://github.com/dotnet/aspnetcore/pull/42901#issuecomment-1260404507 is that dropping support for Type would *not* be a regression because the existing implementation already doesn't work for Type due to a bug. That is, we attempt a comparison like `paramValue.GetType() == typeof(Type)`, but this comparison would always evaluate as `false` because `paramValue.GetType()` will be something like `System.RuntimeType` rather than `System.Type`.

So it would be a free choice about whether or not we want to *add* support for Type, and the argument made by @alix-tlse is that this is an uncommon scenario so we should not add support for that.

> IEventCallback is not immutable.

Yes, very true. Are you saying we should be checking for `EventCallback` instead of `IEventCallback`?

Reply:
Hi, to explain why I chose `IEventCallback`, it is to cover both types `EventCallback` and `EventCallback<T>`. This interface is "internal" and both types are structures with read-only fields. I did some tests and an EventCallback parameter is equal to its previous value in my wasm application between two StateHasChange, maybe this is not the case in server mode... It was not the initial subject of my pull request so I can remove this test, but add it can maybe optimize renderer in some cases)

How I test the EventCallback parameter :
**File "index.razor"**
```
<Sample Name="Sample 1" OnClick='(value)=>Console.WriteLine($"Index value : {value}")' />
<Sample Name="Sample 2" />
<Sample Name="Sample 3" Guid="Guid.NewGuid()" />
<button @onclick="() => { }">TEST</button>
```
**File "Sample.razor"**
<div onclick="DoClick">@Name</div>

```
@code {

    [Parameter]
    public Guid Guid { get; set; }

    [Parameter]
    public string Name { get; set; }

    [Parameter]
    public EventCallback<string> OnClick { get; set; }

    private async Task DoClick()
    {
        if (OnClick.HasDelegate)
        {
            await OnClick.InvokeAsync("Foo");
        }
    }

    private readonly System.Collections.Generic.Dictionary<string, object> _oldParameters = new();

    public override async System.Threading.Tasks.Task SetParametersAsync(ParameterView parameters)
    {
        await base.SetParametersAsync(parameters);
        System.Text.StringBuilder sb = new();
        bool isEqual;
        foreach (ParameterValue par in parameters)
        {
            if (_oldParameters.TryGetValue(par.Name, out object oldValue))
            {
                isEqual = Equals(oldValue, par.Value);
            }
            else
            {
                isEqual = false;
            }
            if (!isEqual)
            {
                sb.AppendLine();
                sb.Append('\t');
                sb.Append($"Name: {par.Name} {(par.Cascading ? "[CASCADING] " : "")}, Value: {par.Value}");
                if (!_oldParameters.TryAdd(par.Name, par.Value))
                {
                    _oldParameters[par.Name] = par.Value;
                }
            }
        }
        Console.WriteLine($"{Name}: {(sb.Length == 0 ? "NO CHANGES" : sb)}");
    }
}
```
**Results :**
After click on the "TEST" button in the browser console :
```
Sample 1: NO CHANGES
Sample 3:  CHANGE : Name: Guid , Value: f3d33409-7873-4f45-8e12-8a929efef691
```
(remarks : The Sample 2 don't call the SetParametersAsync)